### ISSUE TITLE
Concurrent grammar requests

### ISSUE THREAD

**Issue by 28Smiles** [huggingface/text-generation-inference#1643]
### System Info
I am seeing traces of error and crash of the generation process when there is a text generation process in progress and another request that has a definition of json grammar is sent in parallel. I am using TGI from docker. I have tried several models and the problem persists.

The request with json grammar that causes the crash if I send it alone (without any other generation process running) works perfectly and returns the expected structure. Also parallel requests without grammar work without problems.

error stack:
```
2024-03-13 23:39:23 2024-03-13T22:39:23.229128Z ERROR text_generation_launcher: Method Decode encountered an error.
2024-03-13 23:39:23 Traceback (most recent call last):
2024-03-13 23:39:23   File "/opt/conda/bin/text-generation-server", line 8, in <module>
2024-03-13 23:39:23     sys.exit(app())
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/typer/main.py", line 311, in __call__
2024-03-13 23:39:23     return get_command(self)(*args, **kwargs)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
2024-03-13 23:39:23     return self.main(*args, **kwargs)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/typer/core.py", line 778, in main
2024-03-13 23:39:23     return _main(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/typer/core.py", line 216, in _main
2024-03-13 23:39:23     rv = self.invoke(ctx)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
2024-03-13 23:39:23     return _process_result(sub_ctx.command.invoke(sub_ctx))
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
2024-03-13 23:39:23     return ctx.invoke(self.callback, **ctx.params)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 783, in invoke
2024-03-13 23:39:23     return __callback(*args, **kwargs)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/typer/main.py", line 683, in wrapper
2024-03-13 23:39:23     return callback(**use_params)  # type: ignore
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/cli.py", line 89, in serve
2024-03-13 23:39:23     server.serve(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/server.py", line 235, in serve
2024-03-13 23:39:23     asyncio.run(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/asyncio/runners.py", line 44, in run
2024-03-13 23:39:23     return loop.run_until_complete(main)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
2024-03-13 23:39:23     self.run_forever()
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
2024-03-13 23:39:23     self._run_once()
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/asyncio/base_events.py", line 1909, in _run_once
2024-03-13 23:39:23     handle._run()
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/asyncio/events.py", line 80, in _run
2024-03-13 23:39:23     self._context.run(self._callback, *self._args)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/grpc_interceptor/server.py", line 165, in invoke_intercept_method
2024-03-13 23:39:23     return await self.intercept(
2024-03-13 23:39:23 > File "/opt/conda/lib/python3.10/site-packages/text_generation_server/interceptor.py", line 21, in intercept
2024-03-13 23:39:23     return await response
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_aio_server.py", line 82, in _unary_interceptor
2024-03-13 23:39:23     raise error
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_aio_server.py", line 73, in _unary_interceptor
2024-03-13 23:39:23     return await behavior(request_or_iterator, context)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/server.py", line 146, in Decode
2024-03-13 23:39:23     batch = self.model.batch_type.concatenate(batches)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/contextlib.py", line 79, in inner
2024-03-13 23:39:23     return func(*args, **kwds)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/models/flash_causal_lm.py", line 593, in concatenate
2024-03-13 23:39:23     next_token_chooser = HeterogeneousNextTokenChooser.from_pb(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/tokens.py", line 471, in from_pb
2024-03-13 23:39:23     return HeterogeneousNextTokenChooser(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/tokens.py", line 269, in __init__
2024-03-13 23:39:23     HeterogeneousGrammarLogitProcessor(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/logits_process.py", line 558, in __init__
2024-03-13 23:39:23     fsm = GrammarLogitProcessor._cached_compile_fsm(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/logits_process.py", line 518, in _cached_compile_fsm
2024-03-13 23:39:23     fsm = RegexFSM(schema, tokenizer)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/fsm.py", line 121, in __init__
2024-03-13 23:39:23     self.states_to_token_maps, self.empty_token_ids = create_states_mapping(
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/outlines/caching.py", line 74, in wrapper
2024-03-13 23:39:23     result = cached_function(*args, **kwargs)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/fsm.py", line 103, in create_states_mapping
2024-03-13 23:39:23     regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/regex.py", line 165, in make_deterministic_fsm
2024-03-13 23:39:23     new_alphabet = BetterAlphabet(new_symbol_mapping)
2024-03-13 23:39:23   File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/regex.py", line 17, in __init__
2024-03-13 23:39:23     assert anything_else in self._symbol_mapping
2024-03-13 23:39:23 AssertionError
2024-03-13 23:39:23 
2024-03-13 23:39:23 2024-03-13T22:39:23.231722Z ERROR batch{batch_size=2}:decode:decode{size=2}:decode{size=2}: text_generation_client: router/client/src/lib.rs:33: Server error:
```

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

```
import json
import time
import random
import requests
import sys

from concurrent.futures import ThreadPoolExecutor

## read arguments (number of requests to fire)
n = 10
if len(sys.argv) > 1:
    n = int(sys.argv[1])

print("Firing", n, "requests...")

# build the request
json_schema = {
    "properties": {
        "colors": {
            "description": "A list of colors",
            "examples": ["red", "green", "blue"],
            "items": {"type": "string"},
            "title": "Colors",
            "type": "array",
        },
    },
    "required": ["colors"],
}


def get_input():
    return random.choice(
        [
            "the sky",
            "the grass",
            "the sea",
            "the sun",
            "fire",
            "a tomato",
            "a strawberry",
            "a blueberry",
        ]
    )


# prepare function to fire requests
def fire_request():
    try:
        data = {
            "inputs": f"What colors are {get_input()}?",
            "parameters": {
                "best_of": 1,
                "repetition_penalty": 1.3,
                "do_sample": False,
                "grammar": {"type": "json", "value": json_schema},
            },
        }

        headers = {
            "Content-Type": "application/json",
        }

        response = requests.post(
            "http://127.0.0.1:8080/generate", headers=headers, json=data
        )
        # print(response.text)
        response_data = response.json()
        grammar_response = json.loads(response_data["generated_text"])
        print(json.dumps(grammar_response, indent=2))
    except Exception as e:
        # expect that some requests will fail
        print(e, response.text)

def fire_request_no_grammar():
    try:
        data = {
            "inputs": f"What colors are {get_input()}?",
            "parameters": {
                "best_of": 1,
                "repetition_penalty": 1.3,
                "do_sample": False,
            },
        }

        headers = {
            "Content-Type": "application/json",
        }

        response = requests.post(
            "http://127.0.0.1:8080/generate", headers=headers, json=data
        )
        # print(response.text)
        response_data = response.json()
        print(response_data.generated_text, indent=2)
    except Exception as e:
        # expect that some requests will fail
        print(e, response.text)

## fire N requests in parallel
start = time.time()
with ThreadPoolExecutor() as executor:
    for _ in range(n):
        executor.submit(fire_request)
        executor.submit(fire_request_no_grammar)
end = time.time()
print("Time taken:", end - start)
```

### Expected behavior

Working

**Comment by drbh** [comment]
@28Smiles thanks for sharing the reproduction script, I'll take a look shortly and share an update here

**Comment by drbh** [comment]
Hi @28Smiles, I just ran the script above with `gpt2`, `google/gemma-7b-it` and `mistralai/Mistral-7B-Instruct-v0.1` and was unable to reproduce the error. Is it possible the version you're using is not the latest?

**Comment by Aruexi** [comment]
Hi, I can confirm this issue using the latest docker image with `mistralai/Mixtral-8x7B-Instruct-v0.1`

**Comment by Johnno1011** [comment]
I can confirm I am also experiencing the same behaviour using mistralai/Mixtral-8x7B-Instruct-v0.1 as well as mistralai/Mixtral-7B-Instruct-v0.1

To identify the problem I did the following
1. Run a while True over a grammar request, see that there are no issues indefinitely.
2. Make a normal text generation request in a different process 
3. Server error

The container itself does not crash and can complete new text generation requests as normal, but only fails when there are grammar requests being processed.

Below is a dump of the logs from the docker container, for clarity I am using `text-generation-inference:1.4.2` and can confirm no issues with concurrent requests when grammar is not used.

```
2024-03-19T10:10:21.647236Z ERROR text_generation_launcher: Method Decode encountered an error.
Traceback (most recent call last):
  File "/opt/conda/bin/text-generation-server", line 8, in <module>
    sys.exit(app())
  File "/opt/conda/lib/python3.10/site-packages/typer/main.py", line 311, in __call__
    return get_command(self)(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/typer/core.py", line 778, in main
    return _main(
  File "/opt/conda/lib/python3.10/site-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/cli.py", line 89, in serve
    server.serve(
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/server.py", line 235, in serve
    asyncio.run(
  File "/opt/conda/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/conda/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/opt/conda/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/opt/conda/lib/python3.10/asyncio/base_events.py", line 1909, in _run_once
    handle._run()
  File "/opt/conda/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/opt/conda/lib/python3.10/site-packages/grpc_interceptor/server.py", line 165, in invoke_intercept_method
    return await self.intercept(
> File "/opt/conda/lib/python3.10/site-packages/text_generation_server/interceptor.py", line 21, in intercept
    return await response
  File "/opt/conda/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_aio_server.py", line 82, in _unary_interceptor
    raise error
  File "/opt/conda/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_aio_server.py", line 73, in _unary_interceptor
    return await behavior(request_or_iterator, context)
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/server.py", line 146, in Decode
    batch = self.model.batch_type.concatenate(batches)
  File "/opt/conda/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/models/flash_causal_lm.py", line 591, in concatenate
    next_token_chooser = HeterogeneousNextTokenChooser.from_pb(
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/tokens.py", line 470, in from_pb
    return HeterogeneousNextTokenChooser(
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/tokens.py", line 269, in __init__
    HeterogeneousGrammarLogitProcessor(
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/logits_process.py", line 558, in __init__
    fsm = GrammarLogitProcessor._cached_compile_fsm(
  File "/opt/conda/lib/python3.10/site-packages/text_generation_server/utils/logits_process.py", line 518, in _cached_compile_fsm
    fsm = RegexFSM(schema, tokenizer)
  File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/fsm.py", line 121, in __init__
    self.states_to_token_maps, self.empty_token_ids = create_states_mapping(
  File "/opt/conda/lib/python3.10/site-packages/outlines/caching.py", line 74, in wrapper
    result = cached_function(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/fsm.py", line 103, in create_states_mapping
    regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
  File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/regex.py", line 165, in make_deterministic_fsm
    new_alphabet = BetterAlphabet(new_symbol_mapping)
  File "/opt/conda/lib/python3.10/site-packages/outlines/fsm/regex.py", line 17, in __init__
    assert anything_else in self._symbol_mapping
AssertionError

2024-03-19T10:10:21.647995Z ERROR batch{batch_size=2}:decode:decode{size=2}:decode{size=2}: text_generation_client: router/client/src/lib.rs:33: Server error: 
2024-03-19T10:10:21.648196Z ERROR batch{batch_size=2}:decode:decode{size=2}:decode{size=2}: text_generation_client: router/client/src/lib.rs:33: Server error: 
2024-03-19T10:10:21.648243Z ERROR batch{batch_size=2}:decode:decode{size=2}:decode{size=2}: text_generation_client: router/client/src/lib.rs:33: Server error: 
2024-03-19T10:10:21.649512Z ERROR compat_generate{default_return_full_text=true compute_type=Extension(ComputeType("4-nvidia-geforce-rtx-4090"))}:generate{parameters=GenerateParameters { best_of: Some(1), temperature: None, repetition_penalty: Some(1.2), frequency_penalty: None, top_k: None, top_p: None, typical_p: None, do_sample: false, max_new_tokens: Some(1024), return_full_text: Some(false), stop: [], truncate: None, watermark: false, details: true, decoder_input_details: false, seed: None, top_n_tokens: None, grammar: Some(Json(Object {"properties": Object {"age": Object {"title": String("Age"), "type": String("string")}, "name": Object {"title": String("Name"), "type": String("string")}}, "required": Array [String("name"), String("age")], "title": String("Test"), "type": String("object")})) }}:generate:generate_stream:infer:send_error: text_generation_router::infer: router/src/infer.rs:705: Request failed during generation: Server error: 
2024-03-19T10:10:21.649550Z ERROR compat_generate{default_return_full_text=true compute_type=Extension(ComputeType("4-nvidia-geforce-rtx-4090"))}:generate{parameters=GenerateParameters { best_of: Some(1), temperature: None, repetition_penalty: Some(1.2), frequency_penalty: None, top_k: None, top_p: None, typical_p: None, do_sample: false, max_new_tokens: Some(1024), return_full_text: Some(false), stop: [], truncate: None, watermark: false, details: true, decoder_input_details: false, seed: None, top_n_tokens: None, grammar: None }}:generate:generate_stream:infer:send_error: text_generation_router::infer: router/src/infer.rs:705: Request failed during generation: Server error:
```

**Comment by jack-woodcock** [comment]
I can confirm I am also having this issue, same image as people who have posted here.

**Comment by 28Smiles** [comment]
Yes, i used the latest master and latest release, both had the same error

**Comment by drbh** [comment]
hi @Johnno1011 unfortunately the fix for concurrent request was not included in `1.4.2` and was added in commit [343aa7a](https://github.com/huggingface/text-generation-inference/commit/343aa7a1971840df020524dc2b4943ca5b83714a) after `1.4.3`. We'll be publishing `1.4.4` soon and it should resolve this issue.

@jack-woodcock and @28Smiles I am not able to reproduce with the latest changes, any further instructions for reproducing (or best case a new failing test) would be super helpful! Thanks

**Comment by Aruexi** [comment]
I'm running it on GC g2-standard-24 instance with 2 x NVIDIA L4 and following command:
`docker run -d --name tgi --gpus all --restart unless-stopped -e LOG_LEVEL="info,text_generation_router=debug" --shm-size 1g -v /mnt/data:/data ghcr.io/huggingface/text-generation-inference:latest --model-id mistralai/Mixtral-8x7B-Instruct-v0.1 --trust-remote-code --quantize bitsandbytes-nf4 --num-shard 2 --max-total-tokens 8192 --max-input-length 4096 --max-batch-prefill-tokens 4096`

Then reproducing it for me is as simple as using the TGI swagger docs / curl: 
* start one normal generate request 
* while it is being processed make a second request with any grammar
* observe error response of the first normal request (with error logs as the others have posted)

I hope this is helps

**Comment by Johnno1011** [comment]
> hi @Johnno1011 unfortunately the fix for concurrent request was not included in `1.4.2` and was added in commit [343aa7a](https://github.com/huggingface/text-generation-inference/commit/343aa7a1971840df020524dc2b4943ca5b83714a) after `1.4.3`. We'll be publishing `1.4.4` soon and it should resolve this issue.
> 
> @jack-woodcock and @28Smiles I am not able to reproduce with the latest changes, any further instructions for reproducing (or best case a new failing test) would be super helpful! Thanks

Thank you!
I have played with the two new (CUDA) images from https://github.com/huggingface/text-generation-inference/pkgs/container/text-generation-inference but I don't think they have the changes in. An alternative would be to build the image myself from the main branch, right? :)

Look forward to 1.4.4 regardless
